### PR TITLE
[BOLT][DWARF][NFC] Refactor GDB Index into a new file

### PR DESCRIPTION
**Pull request by sayhaan**
Create a new class and file for functions that update GDB index.

### CODE REVIEW

**Comment by ayermolo** [comment]
Add [BOLT][DWARF] to title

**Comment by maksfb** [comment]
> Add [BOLT][DWARF] to title

And NFC somewhere.

File: bolt/include/bolt/Core/GDBIndex.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  BinaryContext &BC;
+
+  /// Entries for GDB Index Types CU List
+  using GDBIndexTUEntryType = std::vector<GDBIndexTUEntry>;
+  GDBIndexTUEntryType GDBIndexTUEntryVector;
+
+public:
+  GDBIndex(BinaryContext &BC) : BC(BC) {}
+
+  /// Adds an GDBIndexTUEntry if .gdb_index section exists.
```
Comment (ayermolo) [inline bolt/include/bolt/Core/GDBIndex.h:43]: Right now it always happens. Maybe in follow up commit let's change it.
Meanwhile change comment to reflect current behavior.

File: bolt/include/bolt/Core/GDBIndex.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
+//===-- bolt/Core/GDBIndex.h - GDB Index support -------*- C++ -*-===//
```
Comment (ayermolo) [inline bolt/include/bolt/Core/GDBIndex.h:1]: can you format this comment section to 80 characters width

File: bolt/include/bolt/Core/GDBIndex.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+  /// Adds an GDBIndexTUEntry if .gdb_index section exists.
+  void addGDBTypeUnitEntry(const GDBIndexTUEntry &Entry);
+
+  /// Rewrite .gdb_index section if present.
+  void updateGdbIndexSection(
+      CUOffsetMap &CUMap, uint32_t NumCUs,
+      std::unique_ptr<DebugARangesSectionWriter> &ARangesSectionWriter);
+
+  /// Returns all entries needed for Types CU list
```
Comment (ayermolo) [inline bolt/include/bolt/Core/GDBIndex.h:51]: Period

File: bolt/lib/Core/GDBIndex.cpp [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "bolt/Core/GDBIndex.h"
+
+using namespace llvm::bolt;
+using namespace llvm::support::endian;
+
+void GDBIndex::addGDBTypeUnitEntry(const GDBIndexTUEntry &Entry) {
```
Comment (ayermolo) [inline bolt/lib/Core/GDBIndex.cpp:14]: Look like some code that checks the section got lost in refactor.

File: bolt/lib/Core/GDBIndex.cpp [hunk 5]
```diff
@@ -0,0 +1,182 @@
+//===- bolt/Rewrite/GDBIndex.cpp -------------------------------------===//
```
Comment (ayermolo) [inline bolt/lib/Core/GDBIndex.cpp:1]: This isn't in Rewrite

File: bolt/include/bolt/Core/GDBIndex.h [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  BinaryContext &BC;
+
+  /// Entries for GDB Index Types CU List.
+  using GDBIndexTUEntryType = std::vector<GDBIndexTUEntry>;
+  GDBIndexTUEntryType GDBIndexTUEntryVector;
+
+public:
+  GDBIndex(BinaryContext &BC) : BC(BC) {}
+
+  std::mutex DWARFRewriterMutex;
```
Comment (ayermolo) [inline bolt/include/bolt/Core/GDBIndex.h:43]: Rename to more relevant variable.

File: bolt/lib/Core/GDBIndex.cpp [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "bolt/Core/GDBIndex.h"
+
+using namespace llvm::bolt;
+using namespace llvm::support::endian;
+
+void GDBIndex::addGDBTypeUnitEntry(const GDBIndexTUEntry &Entry) {
```
Comment (ayermolo) [inline bolt/lib/Core/GDBIndex.cpp:14]: why was this changed to move by reference/push_bcack from &&/emplace_back?

File: bolt/lib/Core/GDBIndex.cpp [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+  std::lock_guard<std::mutex> Lock(GDBIndexMutex);
+  if (!BC.getGdbIndexSection())
+    return;
+  GDBIndexTUEntryVector.push_back(Entry);
+}
+
+void GDBIndex::updateGdbIndexSection(
+    CUOffsetMap &CUMap, uint32_t NumCUs,
+    std::unique_ptr<DebugARangesSectionWriter> &ARangesSectionWriter) {
```
Comment (ayermolo) [inline bolt/lib/Core/GDBIndex.cpp:23]: Change this to DebugARangesSectionWriter &ARangesSectionWriter

**Review by ayermolo — APPROVED** [review]
